service_configuration_lib/cached_view.py

```diff
@@ -0,0 +1,185 @@
+    def log_inotify_constants() -> None:
+        log.info(f'max_queued_events: {pyinotify.max_queued_events.value}')
+        log.info(f'max_user_instances: {pyinotify.max_user_instances.value}')
+        log.info(f'max_user_watches: {pyinotify.max_user_watches.value}')
+
+    def process_events(self) -> None:
+        """This method should be periodically invoked to process queued inotify
+        events.  See the class docstring for more context.
+        """
+        while self._notifier.check_events(timeout=0):
+            self._notifier.read_events()
+            self._notifier.process_events()
+
+    def setup(self) -> None:
+        """Recreates state of ConfigsFileWatcher. Used as reaction on queue overflow in _EventHandler"""
+        handler = _EventHandler(cache=self)
+        watch_manager = pyinotify.WatchManager()
+        self._notifier = pyinotify.Notifier(
+            watch_manager=watch_manager,
+            default_proc_fun=handler,
+        )
+
+        watch_manager.add_watch(
+            path=self._configs_folder,
+            mask=pyinotify.IN_MOVED_TO | pyinotify.IN_CREATE | pyinotify.IN_DELETE,
+            rec=True,
+            auto_add=True,
+            exclude_filter=self._exclude_filter,
+        )
+        for root, _, files in os.walk(self._configs_folder):
+            for path in files:
+                self._maybe_add_path_to_cache(os.path.join(root, path))
+
+    def _exclude_filter(self, path: str) -> bool:
+        """Directories (not files) to exclude from watching.
+
+        Attempting to watch tmp directories causes pyinotify to log an error,
+        so let's just filter them out.
+        Also it will exclude all folders which don't match with services_names param.
+        """
+        folder_name = Path(path).name
+
+        for exclude_folder in self._exclude_folders_filters:
+            if fnmatch.fnmatch(folder_name, exclude_folder):
+                return True
+
+        for service_name in self._services_names:
+            if fnmatch.fnmatch(folder_name, service_name):
+                return False
+
+        return True
+
+    def _maybe_add_path_to_cache(self, path: str) -> None:
+        result = self._service_name_and_config_from_path(path)
+        if result is None:
+            return
+        service_name, config_name, config_suffix = result
+        log.info(f'{config_name}: Config changed for {service_name}')
+        self.configs_view.add(path, service_name, config_name, config_suffix)
+
+    def _maybe_remove_path_from_cache(self, path: str) -> None:
+        result = self._service_name_and_config_from_path(path)
+        if result is None:
+            return
+        service_name, config_name, config_suffix = result
+        log.info(f'{config_name}: Removing config for {service_name}')
+        self.configs_view.remove(path, service_name, config_name, config_suffix)
+
+    def _service_name_and_config_from_path(self, path: str) -> Optional[Tuple[str, str, str]]:
```

Comment:
The API of returning the config and the suffix separately seems... weird to me. I guess it's not _that_ big of a deal. One option would be to bundle it into its own object/namedtuple would make it simpler. Or you could return a tuple of service and the `Path` object itself. Or you could ignore me 😅